Add an overview of how a transaction gets executed in Ethereum PoS [Fixes #8546] (#8548)
ethereum/ethereum-org-website: src/content/developers/docs/consensus-mechanisms/pos/index.md

Whereas under proof-of-work, the timing of blocks is determined by the mining difficulty, in proof-of-stake, the tempo is fixed. Time in proof-of-stake Ethereum is divided into slots (12 seconds) and epochs (32 slots). One validator is randomly selected to be a block proposer in every slot. This validator is responsible for creating a new block and sending it out to other nodes on the network. Also in every slot, a committee of validators is randomly chosen, whose votes are used to determine the validity of the block being proposed.

## How a Transaction Gets Executed in Ethereum PoS {#transaction-execution-ethereum-pos}

> The following provides an overview of how an account's transaction gets executed in Ethereum Proof of Stake, in contrast to [how transactions were mined](/developers/docs/consensus-mechanisms/pow/mining/) in Ethereum PoW.

Reviewer: We are int he PoS page, so no need to be explicit about that again here.

```suggestion
The following provides an end-to-end explanation of how a transaction gets executed in Ethereum Proof of Stake.
```

> 1. A user creates and signs a [transaction](/developers/docs/transactions/) with an [account's](/developers/docs/accounts/) private key. The user, usually assisted by their wallet, configures the transaction fees that they are prepared to pay in order to include the transaction in a block. These fees go partially to the validator ([tips](/developers/docs/gas/#priority-fee)) or [get burned](/developers/docs/intro-to-ether/#burning-ether) ([base fee](/developers/docs/gas/#base-fee)).

Reviewer: I think we can be a bit more precise here
```suggestion
1. A user creates and signs a [transaction](/developers/docs/transactions/) with their private key. This is usually handled by a wallet or a library such as [ether.js](https://docs.ethers.io/v5/), [web3js](https://web3js.readthedocs.io/en/v1.8.1/), [web3py](https://web3py.readthedocs.io/en/v5/) etc but under the hood the user is making a request to a node using the Ethereum [JSON-RPC API](/developers/docs/apis/json-rpc/). The user defines the amount of gas that they are prepared to pay as a tip to a validator to encourage them to include the transaction in a block. The [tips](/developers/docs/gas/#priority-fee) get paid to the validator while the [base fee](/developers/docs/gas/#base-fee) gets burned.
```

> 2. The wallet sends the transaction to an Ethereum [execution client](/developers/docs/nodes-and-clients/) which verifies its validity.

Reviewer: It doesn't have to originate from a wallet - it could be a user interacting directly with a node, coding in a dev framework, using a web app etc

```suggestion
2. The transaction is submitted to an Ethereum [execution client](/developers/docs/nodes-and-clients/) which verifies its validity. This means ensuring that the sender has enough ETH to fulfil the transaction and they have signed it with the correct key.
```

> 3. If the transaction is valid, the execution client adds it to its mempool and broadcasts it to other nodes, making it part of the public mempool. Advanced users may have sent their transaction instead to specialized and extended execution clients that hide the transaction from the public mempool, making it available to specialized block builders such as [Flashbots Auction](https://docs.flashbots.net/flashbots-auction/overview).

Reviewer: I think it's a bit confusing to refer to the public mempool as there isn't really such a thing as a single coherent public transaction pool - it's just the collection of all the local mempools that each contain the transactions each node has heard about on the gossip network. 

```suggestion
3. If the transaction is valid, the execution client adds it to its local mempool (list of pending transactions) and also broadcasts it to other nodes over the execution layer gossip network. When other nodes hear about the transaction they add it to their local mempool too. Advanced users might refrain from broadcasting their transaction and insteads forward it to specialized block builders such as [Flashbots Auction](https://docs.flashbots.net/flashbots-auction/overview). This allows them to organize the transactions in upcoming blocks for maximum profit (MEV).
```

> 4. The validator client, that has randomly been chosen as the block proposer for the current slot, proposes the next block that defines the new state for the network and broadcasts it to the network. The block proposer can freely choose which valid transactions to include from their local mempool into the block, or can ask an external block builder through a middleware service such as [MEV-Boost](https://docs.flashbots.net/flashbots-mev-boost/introduction) to build a block for them. The communication between the execution and consensus clients is described in more detail in [Connecting the Consensus and Execution Clients](/developers/docs/networking-layer/#connecting-clients).

Reviewer: Think we can break this down a bit more.
```suggestion
4. One of the nodes on the network is the block proposer for the current slot, having previously been selected pseudo-randomly using RANDAO. This node is responsible for building and broadcasting the next block to be added to the Ethereum blockchain and updating the global state. The node is made up of three parts: an execution client, a consensus client and a validator client. The execution client bundles transactions from the local mempool into an "execution payload" and executes them locally to generate a state change. This information is passed to the consensus client where the execution payload is wrapped as part of a "beacon block" that also contains information about rewards, penalties slashings, attestations etc that enable the network to agree on the sequence of blocks at the head of the chain. The communication between the execution and consensus clients is described in more detail in [Connecting the Consensus and Execution Clients](/developers/docs/networking-layer/#connecting-clients).
```

> 5. The consensus clients from other validators on the network track the newly proposed block and verify it via their execution client. The validators then publicly attest as to whether the block is valid or not. If they accept it, their consensus client adds it to the head to their version of the chain and attest to the current state of the chain (once an epoch).

Reviewer: ```suggestion
5. Other nodes receive the new beacon block on the consensus layer gossip network. They pass it to their execution client where the transactions are re-executed locally to ensure the proposed state change is valid. The validator client then attests that the block is valid and is the logical next block in their view of the chain (meaning it builds on the chain withthe greatest weight of attestations as defined in the fork choice rules). The block is added to the local database in each node that attests to it.
```

> 6. The other consensus clients on the network run a [fork selection algorithm](/developers/docs/consensus-mechanisms/pos/#fork-choice) and add the block to the fork containing blocks with the most attestations from all validators (weighted by staked Ether).

Reviewer: ```suggestion
```
Think this is better wrapped into the previous point

> 7. The transaction can be considered to have "finality", i.e., that it can not be reverted, when the block that it's been included in has become part of a chain with a subsequent finalized checkpoint block. This occurs, under normal circumstances (i.e., if no major validators go offline or major client bugs have occurred) after two epochs, or approximately 12.8 minutes. Although typically, the chain, and therefore our user's transaction, can be assumed as having reached finality with high probability much sooner in practice. This can be assessed by monitoring the attestation data contained in previous the blocks' metadata. More detail on finality can be found in the [next section](/developers/docs/consensus-mechanisms/pos/#finality).

Reviewer: ```suggestion
6. The transaction can be considered "finalized", i.e., that it can not be reverted, if it has become part of a chain with a "supermajority link" between two checkpoints. Checkpoints occur at the start of each epoch and to have a supermajority link they must both be attested to by 66% of the total staked ETH on the network.

More detail on finality can be found below.
```

Whereas under proof-of-work, the timing of blocks is determined by the mining difficulty, in proof-of-stake, the tempo is fixed. Time in proof-of-stake Ethereum is divided into slots (12 seconds) and epochs (32 slots). One validator is randomly selected to be a block proposer in every slot. This validator is responsible for creating a new block and sending it out to other nodes on the network. Also in every slot, a committee of validators is randomly chosen, whose votes are used to determine the validity of the block being proposed.

## How a Transaction Gets Executed in Ethereum PoS {#transaction-execution-ethereum-pos}

> The following provides an end-to-end explanation of how a transaction gets executed in Ethereum Proof of Stake.

Reviewer 2: ```suggestion
The following provides an end-to-end explanation of how a transaction gets executed in Ethereum proof-of-stake.
```

updating for style guide

1. A user creates and signs a [transaction](/developers/docs/transactions/) with their private key. This is usually handled by a wallet or a library such as [ether.js](https://docs.ethers.io/v5/), [web3js](https://web3js.readthedocs.io/en/v1.8.1/), [web3py](https://web3py.readthedocs.io/en/v5/) etc but under the hood the user is making a request to a node using the Ethereum [JSON-RPC API](/developers/docs/apis/json-rpc/). The user defines the amount of gas that they are prepared to pay as a tip to a validator to encourage them to include the transaction in a block. The [tips](/developers/docs/gas/#priority-fee) get paid to the validator while the [base fee](/developers/docs/gas/#base-fee) gets burned.

> 2. The transaction is submitted to an Ethereum [execution client](/developers/docs/nodes-and-clients/) which verifies its validity. This means ensuring that the sender has enough ETH to fulfil the transaction and they have signed it with the correct key.

Reviewer 2: ```suggestion
2. The transaction is submitted to an Ethereum [execution client](/developers/docs/nodes-and-clients/#execution-client) which verifies its validity. This means ensuring that the sender has enough ETH to fulfill the transaction and they have signed it with the correct key.
```

3. If the transaction is valid, the execution client adds it to its local mempool (list of pending transactions) and also broadcasts it to other nodes over the execution layer gossip network. When other nodes hear about the transaction they add it to their local mempool too. Advanced users might refrain from broadcasting their transaction and insteads forward it to specialized block builders such as [Flashbots Auction](https://docs.flashbots.net/flashbots-auction/overview). This allows them to organize the transactions in upcoming blocks for maximum profit ([MEV](/developers/docs/mev/#mev-extraction)).

> 4. One of the nodes on the network is the block proposer for the current slot, having previously been selected pseudo-randomly using RANDAO. This node is responsible for building and broadcasting the next block to be added to the Ethereum blockchain and updating the global state. The node is made up of three parts: an execution client, a consensus client and a validator client. The execution client bundles transactions from the local mempool into an "execution payload" and executes them locally to generate a state change. This information is passed to the consensus client where the execution payload is wrapped as part of a "beacon block" that also contains information about rewards, penalties slashings, attestations etc that enable the network to agree on the sequence of blocks at the head of the chain. The communication between the execution and consensus clients is described in more detail in [Connecting the Consensus and Execution Clients](/developers/docs/networking-layer/#connecting-clients).

Reviewer 2: ```suggestion
4. One of the nodes on the network is the block proposer for the current slot, having previously been selected pseudo-randomly using RANDAO. This node is responsible for building and broadcasting the next block to be added to the Ethereum blockchain and updating the global state. The node is made up of three parts: an execution client, a consensus client and a validator client. The execution client bundles transactions from the local mempool into an "execution payload" and executes them locally to generate a state change. This information is passed to the consensus client where the execution payload is wrapped as part of a "beacon block" that also contains information about rewards, penalties, slashings, attestations etc that enable the network to agree on the sequence of blocks at the head of the chain. The communication between the execution and consensus clients is described in more detail in [Connecting the Consensus and Execution Clients](/developers/docs/networking-layer/#connecting-clients).
```
